sway-core/src/semantic_analysis/ast_node/declaration/auto_impl.rs

```diff
@@ -471,404 +396,703 @@ impl<'a, 'b> AutoImplAbiEncodeContext<'a, 'b> {
+
+            match decl.purity {
+                Purity::Pure => {},
+                Purity::Reads => {reads = true},
+                Purity::Writes => {writes = true},
+                Purity::ReadsWrites => {
+                    reads = true;
+                    writes = true;
+                },
+            }
+
+            let args_types = itertools::intersperse(
+                decl.parameters.iter().map(|x| {
+                    self.generate_type(engines, x.type_argument.type_id)
+                }),
+                ", ".into()
+            ).collect::<String>();
+
+            let args_types = if args_types.is_empty() {
+                "()".into()
+            } else {
+                format!("({args_types},)")
+            };
+
+            let expanded_args = itertools::intersperse(
+                decl.parameters.iter()
+                    .enumerate()
+                    .map(|(i, _)| format!("args.{i}")),
+                ", ".into()
+            ).collect::<String>();
+
+            let return_type = self.generate_type(engines, decl.return_type.type_id);
+
+            let method_name = decl.name.as_str();
+
+            code.push_str(&format!("if method_name == \"{method_name}\" {{
+                let args = decode_second_param::<{args_types}>();
+                let result_{method_name}: raw_slice = encode::<{return_type}>(__contract_entry_{method_name}({expanded_args}));
+                __contract_ret(result_{method_name}.ptr(), result_{method_name}.len::<u8>());
+            }}\n"));
         }
+
+        let att: String = match (reads, writes) {
+            (true, true) => "#[storage(read, write)]",
+            (true, false) => "#[storage(read)]",
+            (false, true) => "#[storage(write)]",
+            (false, false) => ""
+        }.into();
+
+        let code = format!("{att}
+        pub fn __entry() {{
+            let method_name = decode_first_param::<str>();
+            {code}
+        }}");
+
+        self.parse_item_fn_to_typed_ast_node(engines, FunctionDeclarationKind::Entry, &code)
+    }
+
+    pub(crate) fn generate_predicate_entry(&mut self, engines: &Engines, decl: &TyFunctionDecl) -> Option<TyAstNode> {
+        let args_types = itertools::intersperse(
+            decl.parameters.iter().map(|x| {
+                self.generate_type(engines, x.type_argument.type_id)
+            }),
+            ", ".into()
+        ).collect::<String>();
+
+        let args_types = if args_types.is_empty() {
+            "()".into()
+        } else {
+            format!("({args_types},)")
+        };
+
+        let expanded_args = itertools::intersperse(
+            decl.parameters.iter()
+                .enumerate()
+                .map(|(i, _)| format!("args.{i}")),
+            ", ".into()
+        ).collect::<String>();
+
+        let code = format!("pub fn __entry() -> bool {{
+            let args = decode_script_data::<{args_types}>();
+            main({expanded_args})
```

Comment:
What will happen if `main` is wrongly declared and does not return `bool`? E.g. `fn main() -> u64`. How will error reporting look like?

Right now, I assume `__entry` will be generated nevertheless and while parsing it we will have an additional type mismatch error in the generated code which will trigger an assert in the `parse_item_fn_to_typed_ast_node`.

Reply:
Currently is correctly only generating one error:

```
Testing should_pass/language/predicate_while ... failed
     Failed to compile predicate_while
          Compiling should_pass/language/predicate_while ...

          Failed to compile predicate_while
          error
           --> /home/xunilrj/github/sway/test/src/e2e_vm_tests/test_programs/should_pass/language/predicate_while/src/main.sw:3:1
            |
          1 |   
          2 |   
          3 | / fn main() -> u64 {
          4 | |     0
          5 | | }
            | |_^ A predicate's main function must return a boolean.
            |
          ____

            Aborting due to 1 error.
```

What happens is that the module `type_check` fails and the code is never generated.